### PR TITLE
ARROW-9147: [C++][Dataset] Support projection from null->any type

### PR DESCRIPTION
**Pull request by bkietz**

### CODE REVIEW

File: cpp/src/arrow/dataset/projector.cc [hunk 1]
```diff
@@ -46,6 +46,15 @@ Status CheckProjectable(const Schema& from, const Schema& to) {
                                from);
     }
 
+    if (from_field->type()->id() == Type::NA) {
+      // promotion from null to any type is supported
+      if (to_field->nullable()) continue;
+
+      return Status::TypeError("field ", to_field->ToString(),
+                               " is not nullable and but has type ", NullType(),
```
Comment (pitrou) [inline cpp/src/arrow/dataset/projector.cc:54]: "and but"?

**Review by jorisvandenbossche — COMMENTED** [review]
Generally looks good to me, thanks! (the related tests are failing, though, on some builds)

File: cpp/src/arrow/dataset/projector.cc (hunk 1 above)
Comment (bkietz) [reply cpp/src/arrow/dataset/projector.cc:54]: oops, copypasta error

File: python/pyarrow/tests/test_dataset.py [hunk 2]
```diff
@@ -2124,6 +2124,23 @@ def test_dataset_project_only_partition_columns(tempdir):
...
+@pytest.mark.pandas
+def test_dataset_project_null_column(tempdir):
+    import pandas as pd
+    df = pd.DataFrame({"col": np.array([None, None, None], dtype='object')})
+
+    f = tempdir / "test_dataset_project_null_column.parquet"
+    df.to_parquet(f, engine="pyarrow")
+
+    import pyarrow as pa
+    import pyarrow.dataset as ds
```
Comment (pitrou) [inline python/pyarrow/tests/test_dataset.py:2137]: These two imports should be unnecessary?

**Comment by jorisvandenbossche** [comment]
Remaining failures look unrelated?

**Comment by pitrou** [comment]
Yes, they are.

**Review by pitrou — APPROVED** [review]
+1, will merge

**Comment by jorisvandenbossche** [comment]
Thanks @bkietz !

**Pull request closed without merging**